File: diskrit-2/graph/graph-coloring/graph.c
#include "graph.h"

typedef struct node
{
    int vertex;
    struct node* next;
} node_t;

typedef struct graph
{
    int vertices;
    struct node** Adjlist;
} graph_t;


node_t *create_nodes(int vertex){
    node_t *new_node = malloc(sizeof(node_t));
    new_node->vertex = vertex;
    new_node->next = NULL;
    return new_node;
}

graph_t *create_graph(int vertices){
    graph_t *new_graph = malloc(sizeof(graph_t));
    new_graph->vertices = vertices;
    new_graph->Adjlist = calloc(sizeof(node_t), new_graph->vertices);

    for (int i = 0; i < vertices; i++)
    {
        new_graph->Adjlist[i] = NULL;
    }
    
    return new_graph;
}

void add_edges(graph_t *graph, int vertex_a, int vertex_b){
    node_t *temp = create_nodes(vertex_b);
    temp->next = graph->Adjlist[vertex_a];
    graph->Adjlist[vertex_a] = temp;
}

void print_graph(graph_t *graph){
    int *degree = calloc(graph->vertices, sizeof(int));
    printf("\nAdjacency List:");
    node_t *temp = NULL;
    for (int i = 0; i < graph->vertices; i++)
    {
        temp = graph->Adjlist[i];
        printf("\nVertex %d: ", i);
        degree[i] = 0;
        
        while (temp)
        {
            printf("%d -> ", temp->vertex);
            temp = temp->next;
            degree[i]++;
        }
        printf("| degree: %d", degree[i]);
    }
}

void dfs(graph_t *graph, int random_start, int *reach){
    reach[random_start] = 1;
    node_t *temp = graph->Adjlist[random_start];
    while (temp)
    {
        if (reach[temp->vertex] != 1)
        {
            dfs(graph, temp->vertex, reach);
        }
        temp = temp->next;
    }
}

bool check_graph(graph_t *graph){
    int count_check = 0;
    node_t *check;
    for (int i = 0; i < graph->vertices; i++)
    {
        check = graph->Adjlist[i];
        if (check == NULL)
        {
            count_check++;
        }
    }

    if (count_check == graph->vertices)
    {
        return false;
    }
    else
    {
        return true;
    }
    
}

void is_connected(graph_t *graph){
    if (!check_graph(graph))
    {
        printf("\nNot a Graph");
        return;
    }
    
    int *reach = calloc(graph->vertices, sizeof(int));
    int count;
    
    for (int i = 0; i < graph->vertices; i++)
    {
        count = 0;
        for (int j = 0; j < graph->vertices; j++)
        {
            reach[j] = 0;
        }
        
        dfs(graph, i, reach);

        for (int j = 0; j < graph->vertices; j++)
        {
            if (reach[j])
            {
                count++;
            }
        }

        if (count == graph->vertices)
        {
            printf("\nGraph is Connected");
            return;
        }
    }
    printf("\nGraph is Disconnected");
}

void is_directed(graph_t *graph){
    if (!check_graph(graph))
    {
        return;
    }
    
    int *reach1 = calloc(graph->vertices, sizeof(int));
    int *reach2 = calloc(graph->vertices, sizeof(int));

    for (int i = 1; i < graph->vertices; i++)
    {
        dfs(graph, i-1, reach1);
        dfs(graph, i, reach2);
        for (int j = 0; j < graph->vertices; j++)
        {
            if ((reach1[j] && !reach2[j]) || (!reach1[j] && reach2[j]))
            {
                printf("\nGraph is Directed");
                return;
            }
        }
        
    }
    printf("\nGraph is Undirected");
}

void greedy_coloring(graph_t *graph){
    if (!check_graph(graph))
    {
        printf("Undefined");
        return;
    }

    int *result = calloc(graph->vertices, sizeof(int));
    bool *available = calloc(graph->vertices, sizeof(bool));

    for (int i = 0; i < graph->vertices; i++) {
        result[i] = -1;
        available[i] = true;
    }

    result[0] = 0;

    // Special case if the user ONLY inputted two vertices
    if (graph->vertices == 2) {
        result[0] = 0;
        result[1] = 1;
        for (int i = 0; i < graph->vertices; i++)
        {
            printf("Vertex %d --->  Color %d\n", i, result[i]);
        }
        return;
    }

    node_t *temp = NULL;
    int count;
    for (int i = 1; i < graph->vertices; i++)
    {
        temp = graph->Adjlist[i];
        while (temp)
        {
            if (result[temp->vertex] != -1)
            {
                available[result[temp->vertex]] = false;
            }
            temp = temp->next;
        }

        for (count = 0; count < graph->vertices; count++){
            if (available[count]){
                break;
            }
        }

        result[i] = count;

        temp = graph->Adjlist[i];
        while (temp)
        {
            if (result[temp->vertex] != -1)
            {
                available[result[temp->vertex]] = true;
            }
            temp = temp->next;
        }
    }

    for (int i = 0; i < graph->vertices; i++)
    {
        printf("Vertex %d --->  Color %d\n", i, result[i]);
    }
}

File: diskrit-2/graph_python/graph_coloring/graph.py
class Node:
    def __init__(self, data=None) -> None:
        self.data = data
        self.next = None
        

class Graph:
    def __init__(self, num = int) -> None:
        self.size = num
        self.graph = [None] * self.size

    def add_edge(self, vertex_a, vertex_b) -> None:
        node = Node(vertex_b)
        node.next = self.graph[vertex_a]
        self.graph[vertex_a] = node

    def print_list(self) -> None:
        print("\nAdjancecy List: \n")
        for i in range(self.size):
            print(f"Vertex {i}:", end="")
            temp = self.graph[i]
            while temp:
                print(f" -> {temp.data}", end="")
                temp = temp.next
            print(" \n")

    def coloring(self) -> None:
        result = [-1] * self.size
        available = [True] * self.size
        if self.size == 2:
            result[0] = 0
            result[1] = 1
            for i in range(self.size):
                print(f"Vertex {i} --> Color {result[i]}")
        else:
            result[0] = 0
            for v in range(1, self.size):
                temp = self.graph[v]
                while temp:
                    if result[temp.data] != -1:
                        available[result[temp.data]] = False
                    temp = temp.next
                for count in range(self.size):
                    if available[count]:
                        break
                result[v] = count
                temp = self.graph[v]
                while temp:
                    if result[temp.data] != -1:
                        available[result[temp.data]] = True
                    temp = temp.next
            for i in range(self.size):
                print(f"Vertex {i} --> Color {result[i]}")

    def dfs(self, random_start = int, reach = []) -> None:
        reach[random_start] = 1
        temp = self.graph[random_start]
        while temp:
            if reach[temp.data] != 1:
                self.dfs(temp.data, reach)
            temp = temp.next
    
    def is_connected(self) -> None:
        for i in range(self.size):
            count = 0
            reach = [0] * self.size
            self.dfs(i, reach)
            for j in reach:
                if j == 1:
                    count += 1
            if count == self.size:
                print("\nGraph is Connected")
                return
        print("\nGraph is Disconected")
        return
    
    def is_directed(self) -> None:
        reach1 = [0] * self.size
        reach2 = [0] * self.size
        for i in range(self.size):
            self.dfs(i - 1, reach1)
            self.dfs(i, reach2)
            for j in range(self.size):
                if (reach1[j] == 1 and reach2[j] == 0) or (reach1[j] == 0 and reach2[j] == 1):
                    print("\nGraph is Directed")
                    return
        print("\nGraph is Undirected")

    